kratos/utilities/assign_mpcs_to_neighbours_utility.cpp

```diff
@@ -0,0 +1,376 @@
+    mpBins = Kratos::make_unique<NodeBinsType>(nodes_model_part.begin(), nodes_model_part.end());
+    mMaxNumberOfNodes = rStructureNodes.size();
+    KRATOS_CATCH(""); 
+} 
+
+// Search for the neighbouring nodes (in rStructureNodes) of each rNode on a given array of rNodes
+void AssignMPCsToNeighboursUtility::SearchNodesInRadiusForNodes(
+    NodesContainerType const& rNodes,
+    const double Radius,
+    VectorResultNodesContainerType& rResults,
+    const double MinNumOfNeighNodes)
+{
+    KRATOS_TRY;
+    NodesContainerType::ContainerType& nodes_array     = const_cast<NodesContainerType::ContainerType&>(rNodes.GetContainer());
+
+    rResults.resize(nodes_array.size());
+
+    // Declare a counter variable outside the loop as std::atomic<int>
+    std::atomic<int> i(0);
+
+    block_for_each(nodes_array, [&](Node<3>::Pointer& pNode)
+    {
+        double localRadius = Radius;
+
+        IndexType it = i.fetch_add(1); // Atomically increment the counter and get the previous value
+
+        while (rResults[it].size()<MinNumOfNeighNodes){
+            rResults[it].clear();
+            SearchNodesInRadiusForNode(pNode, localRadius, rResults[it]);
+            localRadius += Radius;
+        }
+    });
+    KRATOS_CATCH("");
+}
+
+
+//Search for the neighbouring nodes (in rStructureNodes) of the given rNode
+void AssignMPCsToNeighboursUtility::SearchNodesInRadiusForNode(
+    NodeType::Pointer pNode,
+    double const Radius,
+    ResultNodesContainerType& rResults)
+{
+
+    KRATOS_TRY;
+
+    ResultNodesContainerType local_results(mMaxNumberOfNodes);
+    std::size_t num_of_results = 0;
+    auto results_iterator = local_results.begin();
+    num_of_results = mpBins->SearchInRadius(*pNode, Radius, results_iterator, mMaxNumberOfNodes);
+    rResults.insert(rResults.begin(), local_results.begin(), local_results.begin() + num_of_results);
+
+    KRATOS_CATCH("");
+}
+
+const Variable<double>& AssignMPCsToNeighboursUtility::GetComponentVariable(
+    const Variable<array_1d<double, 3>>& rVectorVariable, 
+    const std::size_t ComponentIndex)
+{
+    static const std::array<std::string, 3> component_suffixes = {"_X", "_Y", "_Z"};
+
+    if (ComponentIndex < 3) {
+        std::string component_variable_name = rVectorVariable.Name() + component_suffixes[ComponentIndex];
+        return KratosComponents<Variable<double>>::Get(component_variable_name);
+    } else {
+        KRATOS_ERROR << "Component variable for " << rVectorVariable.Name() << " at index " << ComponentIndex << " not found" << std::endl;
+    }
+}
+
+
+void AssignMPCsToNeighboursUtility::GetDofsAndCoordinatesForNode(
+    NodeType::Pointer pNode,
+    const Variable<double>& rVariable,
+    DofPointerVectorType& rCloud_of_dofs,
+    array_1d<double,3>& rSlave_coordinates
+    )
+{
+    KRATOS_TRY;
+    rCloud_of_dofs.push_back(pNode->pGetDof(rVariable));
+    rSlave_coordinates = pNode->Coordinates();
```

Comment:
I don't see the point on doing a function of two lines. IMO it only makes more complex browsing the code with no advantage at all.

Reply:
I don't get your point on "similarity"... I'd remove it but if you really require it leave it. Up to you.